docs/tpm_lpc_spi_interface.md

```diff
@@ -0,0 +1,141 @@
+    clock cycles) as it wants, except during read of a few registers where it is
+    allowed to insert only one wait cycle. TPM cannot insert more than 1 wait
+    cycle for the following registers: TPM_ACCESS_x TPM_STS_x
+    TPM_INTF_CAPABILITY TPM_INT_ENABLE TPM_INT_STATUS TPM_INT_VECTOR TPM_DID
+    TPM_VID.
+  - TPM **must** insert wait state if Host attempts to write data without
+    checking TPM status register to see if TPM is ready.
+    > For writes, the TPM is required to insert wait states if software
+    > attempts to write data without waiting for the TPM to transition to the
+    > Ready state.
+- Data transmission (see section 7.4.6 for details)
+  - TPM uses SPI mode 0 (CPHA=0, CPOL=0), no other modes are supported.
+  - SPI packet is outlined in Table 48
+  - Data transfer starts from the most significant bit (MSB) and ends on the
+    least significant bit (LSB)
+  ![](images/tpm2_spi_protocol.png)
+  - The first bit transferred determines whether transaction is read or write.
+  - This is followed by 6 bit length from 1 byte to 64 bytes. This is length of
+    the data payload, excludiong 32-bit header.
+  - Length is followed by 24-bit register address, the highest byte is always
+    0xD4. See Table 9 for details how TPM MMIO space maps to register address
+    emitted on the bus.
+    ![](images/tpm2_locality_address.png)
+
+## TPM register interface
+
+According to section 6.3.1 Host may attempt to read less bytes than there
+actually is in register, TPM must handle this, it must also handle read from the
+middle of register, e.g. if we have a 32-bit register at 0x20, 1 byte read from
+0x20 must yield the first byte of that register, read from 0x21 must yield the
+second byte from that register.
+
+Software must never attempt to cross register size in a same transfer. In case
+of LPC this won't be noticed by TPM as all multi-byte LPC transfers are split
+into multiple transfers, e.g. if hosts writes 0x11223344 to register 0x10 on LPC
+it will appear as
+- 0x11 write to 0x10
+- 0x22 write to 0x11
+- 0x33 write to 0x12
+- 0x44 write to 0x13
+
+Even if register boundary were crossed this doesn't make any difference for TPM.
+In case of SPI data payload of a single transfer cannot be bigger than size of
+register and cannot cross register boundary.
+TBD: how should this be handled? TPM specification says:
+> For SPI, if a TPM receives an access request with a length that exceeds the
+> size of the register specified in the transaction address:
+> (...)
+> b. Writes:
+>   The TPM SHALL update the register designated by the start address.
+>   The TPM MAY update additional, adjacent registers within the targeted
+>   address range.
+>   The TPM SHALL NOT change the state of adjacent registers if the writes to
+>   that register are dropped.
+>
+> The TPM SHOULD NOT abort (as defined in Section 6.5.1.1 Bus Aborts) an entire
+> transaction that crosses a register boundary.
```

Comment:
@krystian-hebel FWIU TPM must stop write on the first error - if write to some register write got dropped it must stop there and don't update further registers. I'm not sure whether TPM should bus abort here or not.

IMO opinion we should try to update as many registers as possible, this would be the closest behaviour to LPC where multi-byte transfers are split into multiple 1 byte transfers.

Reply:
Such approach seems good.